pkg/metrics/metrics.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type OpenstackPrometheusMetrics struct {
+	Duration *prometheus.HistogramVec
+	Total    *prometheus.CounterVec
+	Errors   *prometheus.CounterVec
+}
+
+// MetricPrometheusContext indicates the context for OpenStack metrics.
+type MetricPrometheusContext struct {
+	Start      time.Time
+	Attributes []string
+	Metrics    *OpenstackPrometheusMetrics
+}
+
+// NewMetricPrometheusContext creates a new MetricContext.
+func NewMetricPrometheusContext(resource string, request string) *MetricPrometheusContext {
+	return &MetricPrometheusContext{
+		Start:      time.Now(),
+		Attributes: []string{resource + "_" + request},
+	}
+}
+
+// ObserveRequest records the request latency and counts the errors.
+func (mc *MetricPrometheusContext) ObserveRequest(err error) error {
+	return mc.Observe(APIRequestPrometheusMetrics, err)
+}
+
+// Observe records the request latency and counts the errors.
+func (mc *MetricPrometheusContext) Observe(om *OpenstackPrometheusMetrics, err error) error {
+	if om == nil {
+		// mc.RequestMetrics not set, ignore this request
+		return nil
```

Comment:
Should we maybe return err here, otherwise errors would also be ignored.
```suggestion
		return err
```